### ISSUE TITLE
Insufficient quoting on JoinTable query

### ISSUE THREAD

**Issue by vados-p7e** [typeorm/typeorm#1668]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `websql`

**TypeORM version:**

[ ] `latest`
[x] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

1. Create two entities that don't follow the naming convention (ex. `Thing` in table `things` and `Tag` in table `tags`)

2. Create a join table for them (`thing_tags`) (note that due to the departure from naming convention earlier, this is also a departure from naming convention)

3. Create a `@ManyToMany` relation using `@JoinTable` inside `Thing` as follows:

```ts
  @ManyToMany(type => Tag, {eager: true})
  @JoinTable({
    name: "thing_tags",
    joinColumn: {name: "thing_id", referencedColumnName: "id"}
    inverseJoinColumn: {name: "tag_id", referencedColumnName: "id"}
  })
  tags: Tag[];
```
When trying to retrieve a `Thing`, the query is just barely wrong:

```sql
SELECT Thing_tags_relation_id.tag_id AS "tag_id", Thing_tags_relation_id.thing_id AS "thing_id" FROM "tags" "tags" INNER JOIN "thing_tags" "Thing_tags_relation_id" ON (Thing_tags_relation_id.thing_id = $1 AND Thing_tags_relation_id.tag_id = "tags"."id") ORDER BY Thing_tags_relation_id.tag_id ASC, Thing_tags_relation_id.thing_id ASC;
ERROR:  missing FROM-clause entry for table "thing_tags_relation_id"
```
The problem here is that `Thing_tags_relation_id` is not quoted everywhere it's mentioned. If I take the query to postgres, and add quotes around every unquoted `Thing_tags_relation_id`, the query works great:

```sql
db_test=# SELECT "Thing_tags_relation_id".tag_id AS "tag_id", "Thing_tags_relation_id".thing_id AS "thing_id" FROM "tags" "tags" INNER JOIN "thing_tags" "Thing_tags_relation_id" ON ("Thing_tags_relation_id".thing_id = 1743 AND "Thing_tags_relation_id".tag_id = "tags". "id") ORDER BY "Thing_tags_relation_id".tag_id ASC, "Thing_tags_relation_id".thing_id ASC;
tag_id | thing_id
--------+-----------
(0 rows)
```

**PS** I personally would love a `parametrizedQuery` parameter in error reports that prints the complete query with params filled out (exactly as it would have been sent).

As always thanks for all the work on `typeorm`, it is 100% improving the development lives of me and my team :)

**Comment by pleerock** [comment]
This sounds like a bug, can you please create a PR with a failing test so I can fix this issue quicker ?

**Comment by vados-p7e** [comment]
@pleerock will do, by the end of the week/end

**Comment by pleerock** [comment]
Closing as no answer from reporter.

**Comment by sam-gronblom-rj** [comment]
I feel like I might have run into the same issue. My generated query is missing double quotes around table name and column name for the `ON` part of a `LEFT JOIN`. Specifically:

`FROM "site_service" "SiteService" LEFT JOIN "item_master" "SiteService__item" ON SiteService__
                           item.itemId="SiteService"."itemId"`

This fails in Postgres with the error message: "missing FROM-clause entry for table "siteservice__item""

**Comment by sam-gronblom-rj** [comment]
Is there some easy way to generate isolated test cases?

**Comment by sgronblo** [comment]
Ok, I have a failing test case now. I will see a bit about minimizing the test case.

**Comment by sgronblo** [comment]
Added a new PR which contains a test case which demonstrates a problem that seems similar to this case: #4800

**Comment by sgronblo** [comment]
I dug around the source code and stepped through some execution in a debugger, but still couldn't solve the problem. It seems that the `joinAttributes` of the select query builder's `expressionMap` gets messed up. Specifically it seems that Entity C's "aId" column has itself as its own "referencedColumn". I think I managed to trace it down `join` and `new JoinAttribute`.

**Comment by sgronblo** [comment]
Ok, after digging around some more I realized why I am getting this incorrect query. It's basically because typeorm assumes that a single column will always only be used to reference a single table with a specific referenced column. If you now have a `fooId` relation that refers to `foo.id` and `somefoolinktable.foo_id` by setting up custom JoinColumn decorations with referencedColumnNames you get an incorrect query because one relation will override the referencedColumn of `fooId` that was previously written by the other.

I'm not sure whether this actually is what the original issue poster was experiencing through another pathway.

Does it sound like it would make sense to change the referencedColumns of ColumnMetadata to be `ColumnMetadata` Array instead of just single metadata? I think I will try my hand at making this change but just wanted to check if someone already knew whether it makes sense.

**Comment by rebolyte** [comment]
Just ran into this as well. @pleerock could we get this reopened?

**Comment by bianchi** [comment]
Just ran into this

**Comment by imnotjames** [comment]
Poking at this.  Looks like the it's referencing the wrong value in the join that's created for the relation.

For `C__a` it's creating the query condition of `C__a.barId` when only `C__a.id` is technically possible - so [replacePropertyNames](https://github.com/typeorm/typeorm/blob/0280cdc451c35ef73c830eb1191c95d34f6ce06e/src/query-builder/QueryBuilder.ts#L567) can't correctly escape & translate them from prop to a database name.

**Comment by imnotjames** [comment]
https://github.com/typeorm/typeorm/blob/ec3be4115faba62c1dbbc786ec9283438362bb50/src/metadata-builder/EntityMetadataBuilder.ts#L129

When building the entity metadata the relation is correct when it's getting set here but.. later it's getting over-written.

https://github.com/typeorm/typeorm/blob/ee61c518d746139babbbeaa19b8f3d6e87e34f08/src/metadata-builder/RelationJoinColumnBuilder.ts#L168

We may not be able to store the `referencedColumn` in this way because a single column may have multiple referenced columns from multiple relationships.

Per above, one solution is to make this an array.  We could do that & break the API BC promise.  That.. might be ok in this case.  We could potentially have a getter that's deprecated & points at the last assigned referenced column, though?

Then a list of referenced columns & a helper to fetch the right referencd column given a target.  Could there could be multiple, though??  Should we move this instead off this model and somewhere else?

There's 108 uses of `referencedColumn`.  This could be causing a lot of issues.

**Comment by imnotjames** [comment]
`ColumnMetadata` getting added is helpful but using it for storing relational information is kind of hectic..

**Comment by imnotjames** [comment]
Was thinking it through finding all the related issues.

Path forward can be as such -

* `referencedColumn` has to be marked deprecated. No matter what, this just doesn't work, the data modelling doesn't match up.
* Created a Map or object on the column and map from the relation property name to the referenced column.  Goes from target of current column to the relations property.
* Whenever we need a referenced column we will need to specify for WHAT we need it from - or explicitly pick one at random.  No more implicit randomness here.
* Whatever the last set mapped column is, we set that to the deprecated reference column for now.
* Add a helper in the relation metadata to emit tuples of joins & references.
* Refactor all the code to use one of the multiple mechanisms available to use the correct reference.

**Comment by imnotjames** [comment]
Started implementing the above.  Seems to fix this specific problem - running against the test suite to verify it doesn't break any other cases.

There's a number of places that require refactoring, though, because they use `referencedColumn` and don't know the relation they're associated with.

**Comment by psibean** [comment]
I have an issue with this one coming from QueryBuilder#replacePropertyNames. A query like this:
```
    const myInstance= getRepository().createQueryBuilder()
                         .leftJoin("MainTable", "my_table_relation")
                         .where("MainTable.id = :main_id", { main_id: mainId })
                         .andWhere("my_table_relation.other_id = :other_id", { other_id: ohterId})
                         .getOne();
```
Generates the following SQL:
```
SELECT "MainTable"."id" AS "MainTable_id", "MainTable"."name" AS "MainTable_name", FROM "MainTable" "MainTable" LEFT JOIN "my_table_relation" "my_table_relation" WHERE "MainTable"."id" = $1 AND my_table_relation.other_id = $2

```
I put a debug on the `replacePropertyNames`, for the first run, replacePropertyNames returns the value as `"MainTable"."id"` where as the second time it returns `my_table_relation.other_id` without the quotes as part of the literal string, I think this is an issue with the regex in there or something?

**Comment by nebkat** [comment]
Try use an alias with `createQueryBuilder('main')`, I think it might be mixing things up the entity name is the same as the alias and you are using it in the join also.

**Comment by psibean** [comment]
> Try use an alias with `createQueryBuilder('main')`, I think it might be mixing things up the entity name is the same as the alias and you are using it in the join also.

Hey, I fixed some errors in my example and corrected some statements. I provided an alias as you describe, and this does not fix the issue. There is Entity named "MainTable" and entity named "OtherTable" they share a many-to-many relation, the related table in the database is "main_other_table"
```
QueryFailedError: syntax error at or near "WHERE"
```

Note that the repository being used blow is the repository obtained from MainTable.getRepository():
```
    const mainInstance = mainTableRepository.createQueryBuilder("main")
                         .leftJoin("OtherTable", "main_other_table")
                         .where("main.id = :main_id", { main_id: mainId })
                         .andWhere("OtherTable.other_id = :other_id", { other_id: otherId})
                         .getOne();
```

And this still runs the following SQL:
```
SELECT "main"."id" AS "main_id", "main"."name" AS "main_name" FROM "MainTable" "main" LEFT JOIN "OtherTable" "main_other_table" WHERE "main"."id" = $1 AND main_other_table.other_id = $2
```

However, whilst putting this out here, I've realized in this example that "main_other_table" is being used as an alias here for OtherTable, and isn't specifying that the name of the relational table between them. In this case the where clauses are intended as a filter, not what I wanted them to be joined on, I thought the joins `ON` would have been sorted by identifying the Entity name and the related table name. 

My mistake either way, whether or not the quotation is incorrect due to the query not being validly constructed in the first place, not sure.

**Comment by alfonsoal1983** [comment]
Can you put the code of the entities with the many-to-many relationship?

**Comment by Uzkal** [comment]
> Started implementing the above. Seems to fix this specific problem - running against the test suite to verify it doesn't break any other cases.
> 
> There's a number of places that require refactoring, though, because they use `referencedColumn` and don't know the relation they're associated with.
> 
> In some cases, the code paths are RELYING on this behavior so they don't crash. (They just did the wrong thing before)

Any update on this?

**Comment by khiner** [comment]
Are there any suggested workarounds for this? (Specifically ones using `find`, and not `createQueryBuilder`?)

**Comment by SENETYX** [comment]
Same here,
Whenever I used the join function from a QueryBuilder with the **SAME TABLE**.

```ts
queryBuilder.orderBy('dsr.createdAt', pageOptionsDto.order)
.skip(pageOptionsDto.skip)
.take(pageOptionsDto.take)
.innerJoinAndSelect('dsr.inputs', 'dsr-inputs')
.innerJoinAndSelect('dsr-inputs.file', 'dsr-file') // <-- ❌ remove this and it will work!
.innerJoinAndSelect('dsr-inputs.field', 'dsr-form-field')
.innerJoinAndSelect('dsr-form-field.field', 'dsr-field');
```

Before: **remove the join statement.**

```sql
SELECT 
  "dsr"."id" AS "dsr_id", 
  "dsr"."status" AS "dsr_status", 
  "dsr"."approvedBy" AS "dsr_approvedBy", 
  "dsr"."approvedAt" AS "dsr_approvedAt", 
  "dsr"."createdAt" AS "dsr_createdAt", 
  "dsr"."createdBy" AS "dsr_createdBy", 
  "dsr"."updatedAt" AS "dsr_updatedAt", 
  "dsr"."updatedBy" AS "dsr_updatedBy", 
  "dsr"."dataSubjectId" AS "dsr_dataSubjectId", 
  "dsr"."formId" AS "dsr_formId", 
  "dsr-inputs"."id" AS "dsr-inputs_id", 
  "dsr-inputs"."value" AS "dsr-inputs_value", 
  "dsr-inputs"."createdAt" AS "dsr-inputs_createdAt", 
  "dsr-inputs"."createdBy" AS "dsr-inputs_createdBy", 
  "dsr-inputs"."updatedAt" AS "dsr-inputs_updatedAt", 
  "dsr-inputs"."updatedBy" AS "dsr-inputs_updatedBy", 
  "dsr-inputs"."dsrId" AS "dsr-inputs_dsrId", 
  "dsr-inputs"."fieldId" AS "dsr-inputs_fieldId", 
  "dsr-file"."id" AS "dsr-file_id", 
  "dsr-file"."name" AS "dsr-file_name", 
  "dsr-file"."key" AS "dsr-file_key", 
  "dsr-file"."type" AS "dsr-file_type", 
  "dsr-file"."size" AS "dsr-file_size", 
  "dsr-file"."createdAt" AS "dsr-file_createdAt", 
  "dsr-file"."createdBy" AS "dsr-file_createdBy", 
  "dsr-file"."updatedAt" AS "dsr-file_updatedAt", 
  "dsr-file"."updatedBy" AS "dsr-file_updatedBy", 
  "dsr-form-field"."id" AS "dsr-form-field_id", 
  "dsr-form-field"."isRequired" AS "dsr-form-field_isRequired", 
  "dsr-form-field"."index" AS "dsr-form-field_index", 
  "dsr-form-field"."createdAt" AS "dsr-form-field_createdAt", 
  "dsr-form-field"."createdBy" AS "dsr-form-field_createdBy", 
  "dsr-form-field"."updatedAt" AS "dsr-form-field_updatedAt", 
  "dsr-form-field"."updatedBy" AS "dsr-form-field_updatedBy", 
  "dsr-form-field"."formId" AS "dsr-form-field_formId", 
  "dsr-form-field"."fieldId" AS "dsr-form-field_fieldId", 
  "dsr-field"."id" AS "dsr-field_id", 
  "dsr-field"."status" AS "dsr-field_status", 
  "dsr-field"."fieldDescription" AS "dsr-field_fieldDescription", 
  "dsr-field"."isEncrypt" AS "dsr-field_isEncrypt", 
  "dsr-field"."fieldType" AS "dsr-field_fieldType", 
  "dsr-field"."createdAt" AS "dsr-field_createdAt", 
  "dsr-field"."createdBy" AS "dsr-field_createdBy", 
  "dsr-field"."updatedAt" AS "dsr-field_updatedAt", 
  "dsr-field"."updatedBy" AS "dsr-field_updatedBy" 
FROM 
  "dsr" "dsr" 
  INNER JOIN "dsr-input" "dsr-inputs" ON "dsr-inputs"."dsrId" = "dsr"."id" 
  INNER JOIN "dsr-file" "dsr-file" ON <-- ❌ missing 'ON' statement.
  INNER JOIN "dsr_form_field" "dsr-form-field" ON "dsr-form-field"."id" = "dsr-inputs"."fieldId" 
  INNER JOIN "dsr_field" "dsr-field" ON "dsr-field"."id" = "dsr-form-field"."fieldId" 
ORDER BY 
  "dsr"."createdAt" ASC
```
After: **remove the join statement.**

```sql
SELECT 
  "dsr"."id" AS "dsr_id", 
  "dsr"."status" AS "dsr_status", 
  "dsr"."approvedBy" AS "dsr_approvedBy", 
  "dsr"."approvedAt" AS "dsr_approvedAt", 
  "dsr"."createdAt" AS "dsr_createdAt", 
  "dsr"."createdBy" AS "dsr_createdBy", 
  "dsr"."updatedAt" AS "dsr_updatedAt", 
  "dsr"."updatedBy" AS "dsr_updatedBy", 
  "dsr"."dataSubjectId" AS "dsr_dataSubjectId", 
  "dsr"."formId" AS "dsr_formId", 
  "dsr-inputs"."id" AS "dsr-inputs_id", 
  "dsr-inputs"."value" AS "dsr-inputs_value", 
  "dsr-inputs"."createdAt" AS "dsr-inputs_createdAt", 
  "dsr-inputs"."createdBy" AS "dsr-inputs_createdBy", 
  "dsr-inputs"."updatedAt" AS "dsr-inputs_updatedAt", 
  "dsr-inputs"."updatedBy" AS "dsr-inputs_updatedBy", 
  "dsr-inputs"."dsrId" AS "dsr-inputs_dsrId", 
  "dsr-inputs"."fieldId" AS "dsr-inputs_fieldId", 
  "dsr-form-field"."id" AS "dsr-form-field_id", 
  "dsr-form-field"."isRequired" AS "dsr-form-field_isRequired", 
  "dsr-form-field"."index" AS "dsr-form-field_index", 
  "dsr-form-field"."createdAt" AS "dsr-form-field_createdAt", 
  "dsr-form-field"."createdBy" AS "dsr-form-field_createdBy", 
  "dsr-form-field"."updatedAt" AS "dsr-form-field_updatedAt", 
  "dsr-form-field"."updatedBy" AS "dsr-form-field_updatedBy", 
  "dsr-form-field"."formId" AS "dsr-form-field_formId", 
  "dsr-form-field"."fieldId" AS "dsr-form-field_fieldId", 
  "dsr-field"."id" AS "dsr-field_id", 
  "dsr-field"."status" AS "dsr-field_status", 
  "dsr-field"."fieldDescription" AS "dsr-field_fieldDescription", 
  "dsr-field"."isEncrypt" AS "dsr-field_isEncrypt", 
  "dsr-field"."fieldType" AS "dsr-field_fieldType", 
  "dsr-field"."createdAt" AS "dsr-field_createdAt", 
  "dsr-field"."createdBy" AS "dsr-field_createdBy", 
  "dsr-field"."updatedAt" AS "dsr-field_updatedAt", 
  "dsr-field"."updatedBy" AS "dsr-field_updatedBy" 
FROM 
  "dsr" "dsr" 
  INNER JOIN "dsr_input" "dsr-inputs" ON "dsr-inputs"."dsrId" = "dsr"."id" 
  INNER JOIN "dsr_form_field" "dsr-form-field" ON "dsr-form-field"."id" = "dsr-inputs"."fieldId" 
  INNER JOIN "dsr_field" "dsr-field" ON "dsr-field"."id" = "dsr-form-field"."fieldId" 
ORDER BY 
  "dsr"."createdAt" ASC
```

**Comment by ns-danielxu** [comment]
Any updates on this issue? Facing similar problem

**Comment by MoacirCarlos** [comment]
Any solution? It took me hours to figure out that mapping relationships to different entities using the same source attribute, it got the JOINs mixed up.

took the referencedColumnName from a ManyToOne and put it in another...

I found many open issues on this subject, and no solution.

```
@ManyToOne(() => EntityOne, (e) => e.entitiesOne)
@JoinColumn([
    {
      name: 'co_process',
      referencedColumnName: 'coProcess',
    },
    { name: 'co_center', referencedColumnName: 'coCenter' },
])
process: EntityOne;

@ManyToOne(() => EntityTwo, (e) => e.entitiesExamples)
@JoinColumn([{ name: 'co_center', referencedColumnName: 'coSeqCenter' }])
center: EntityTwo;
```

In my case, for example, he took the "coSeqCenter" and put it in the JOIN of the relationship with the entity "EntityOne".

**Comment by EdgarAlonsoPerez** [comment]
Hi im having the same problem,  it seems that typeOrm is getting issues if the @joinColumn has the same name.
Here's my entities:

```@Entity()
export class CustomAttributeObjectEntity {
 @PrimaryColumn()
  ID: number;

  @Column({ name: 'ATTRIBUTE_ID' })
  ATTRIBUTE_ID: number;

  @Column({ name: 'OBJECT_ID' })
  OBJECT_ID: number;

  @ManyToOne(() => OgObjectEntity, (og) => og.CUSTOM_ATTRIBUTES)
  @JoinColumn({
    name: 'OBJECT_ID',
    referencedColumnName: 'OG_ID',
  })
  OG: OgObjectEntity

  @ManyToOne(() => ComponentEntity, (component) => component.CUSTOM_ATTRIBUTES)
  @JoinColumn({
    name: 'OBJECT_ID',
    referencedColumnName:'COMPONENT_ID',
  })
  COMPONENT: ComponentEntity;
}


@Entity()
export abstract class OgObjectEntity {
  @PrimaryColumn({ name: 'OG_ID'})
  OG_ID: number;

  @OneToMany(() => CustomAttributeObjectEntity, (cao) => cao.OG)
  @JoinColumn({
    name: 'OG_ID',
    referencedColumnName:'OBJECT_ID'
  })
  CUSTOM_ATTRIBUTES: CustomAttributeObjectEntity[]
}



@Entity()
@TableInheritance()
export class ComponentEntity {
  @PrimaryColumn({
    name: 'COMPONENT_ID'
  })
  COMPONENT_ID: number;

  @Column({
    name: 'COMPONENT_NAME'
  })
  COMPONENT_NAME: string;

  @OneToMany(() => CustomAttributeObjectEntity, (customAttribute) => customAttribute.COMPONENT)
  @JoinColumn({
    name: 'COMPONENT_ID',
    referencedColumnName:'OBJECT_ID'
  })
  CUSTOM_ATTRIBUTES: CustomAttributeObjectEntity[]
}
```

**Comment by EdgarAlonsoPerez** [comment]
i solved creating a different classes, here's the same example:

```

@Entity()
export class CustomAttributeObjectEntity {
  @PrimaryColumn()
  ID: number;

  @Column({ name: 'ATTRIBUTE_ID' })
  ATTRIBUTE_ID: number;

  @Column({ name: 'OBJECT_ID' })
  OBJECT_ID: number;  
}

@Entity()
export class ComponentCustomAttributeObjectEntity extends CustomAttributeObjectEntity {

  @Column({ name: 'OBJECT_ID' })
  OBJECT_ID: number;

  @ManyToOne(() => ComponentEntity, (component) => component.CUSTOM_ATTRIBUTES)
  @JoinColumn({
    name: 'OBJECT_ID',
    referencedColumnName: 'COMPONENT_ID',
  })
  COMPONENT: ComponentEntity;
}

@Entity()
export class ServiceCustomAttributeObjectEntity extends CustomAttributeObjectEntity {

  @Column({ name: 'OBJECT_ID' })
  OBJECT_ID: number;

  @ManyToOne(() => OgObjectEntity, (og) => og.CUSTOM_ATTRIBUTES)
  @JoinColumn({
    name: 'OBJECT_ID',
    referencedColumnName: 'OG_ID',
  })
  OG:  ServiceEntity
}
```

**Comment by gh0stek** [comment]
I also had the same problem with relations that shares the same column using `@JoinColumn`.  I noticed that entity keeps the join columns in one collection from all relations and distinguishes it comparing `name` property. So I tried to make `joinColumn1.name !== joinColumn2.name` and I come out with this solution: 
```
@Entity()
export class Item {
  @PrimaryGeneratedColumn('uuid')
  id: string

  @Column()
  name: string

  @Column({
    select: false, // needed to hide the column from the query, because typeorm selects it twice
  })
  item_group_id: number

  @Column({ nullable: true })
  item_sub_group_id?: number

  @ManyToOne(() => ItemGroup)
  @JoinColumn({
    name: new String('item_group_id') as string, // workaround for typeorm issue
    referencedColumnName: 'id',
  })
  itemGroup: ItemGroup

  @ManyToOne(() => ItemSubGroup)
  @JoinColumn([
    {
      name: 'item_group_id',
      referencedColumnName: 'item_group_id',
    },
    {
      name: 'item_sub_group_id',
      referencedColumnName: 'id',
    },
  ])
  itemSubGroup: ItemSubGroup
}
```

This worked pretty well for me. The only issue with this solution is that typeorm thinks that it has two columns named `item_group_id` and adds it twice in select queries but `{ select: false  }` solves this issue. Another problem that it wants to create this column twice when you generate migration but I don't have workaround for that.

**Comment by MoacirCarlos** [comment]
> I also had the same problem with relations that shares the same column using `@JoinColumn`. I noticed that entity keeps the join columns in one collection from all relations and distinguishes it comparing `name` property. So I tried to make `joinColumn1.name !== joinColumn2.name` and I come out with this solution:
> 
> ```
> @Entity()
> export class Item {
>   @PrimaryGeneratedColumn('uuid')
>   id: string
> 
>   @Column()
>   name: string
> 
>   @Column({
>     select: false, // needed to hide the column from the query, because typeorm selects it twice
>   })
>   item_group_id: number
> 
>   @Column({ nullable: true })
>   item_sub_group_id?: number
> 
>   @ManyToOne(() => ItemGroup)
>   @JoinColumn({
>     name: new String('item_group_id') as string, // workaround for typeorm issue
>     referencedColumnName: 'id',
>   })
>   itemGroup: ItemGroup
> 
>   @ManyToOne(() => ItemSubGroup)
>   @JoinColumn([
>     {
>       name: 'item_group_id',
>       referencedColumnName: 'item_group_id',
>     },
>     {
>       name: 'item_sub_group_id',
>       referencedColumnName: 'id',
>     },
>   ])
>   itemSubGroup: ItemSubGroup
> }
> ```
> 
> This worked pretty well for me. The only issue with this solution is that typeorm thinks that it has two columns named `item_group_id` and adds it twice in select queries but `{ select: false }` solves this issue. Another problem that it wants to create this column twice when you generate migration but I don't have workaround for that.

Here I have solved it with "leftJoinAndMapMany/leftJoinAndMapOne" (queryBuilder), with conditionals.

**Comment by yevhen-komarov** [comment]
> Added a new PR which contains a test case which demonstrates a problem that seems similar to this case: [#4800](https://github.com/typeorm/typeorm/pull/4800)

If you need to fix this issue or want to test the solution, install [#11400](https://github.com/typeorm/typeorm/pull/11400#issuecomment-2849034549)

**Comment by yevhen-komarov** [comment]
@EdgarAlonsoPerez 
> i solved creating a different classes, here's the same example:

I solved this problem in a similar way before https://github.com/typeorm/typeorm/issues/11109#issuecomment-2438711505

now the problem is solved, 
If you need to fix this issue or want to test the solution, install [#11400](https://github.com/typeorm/typeorm/pull/11400#issuecomment-2849034549)
